File: Cpp/593_Valid_Square.cpp
class Solution {
public:
    bool validSquare(vector<int>& p1, vector<int>& p2, vector<int>& p3, vector<int>& p4) {
        unordered_map<int, int> m;
        ++m[getDistance(p1, p2)];
        ++m[getDistance(p1, p3)];
        ++m[getDistance(p1, p4)];
        ++m[getDistance(p2, p3)];
        ++m[getDistance(p2, p4)];
        ++m[getDistance(p3, p4)];
        
        if(m.size() != 2) return false;
        for(auto& it : m) {
            if(it.first == 0) return false;
            if(it.second == 4) return true;
        }
        
        return false;
    }
    
    int getDistance(vector<int> a, vector<int> b) {
        int x = a[0] - b[0];
        int y = a[1] - b[1];
        return x * x + y * y; 
    }
};

File: Kotlin/593_Valid_Square.kt
class Solution {
    fun validSquare(p1: IntArray, p2: IntArray, p3: IntArray, p4: IntArray): Boolean {
        val list = mutableListOf<Int>()
        
        list.add(getDistance(p1, p2))
        list.add(getDistance(p1, p3))
        list.add(getDistance(p1, p4))
        list.add(getDistance(p2, p3))
        list.add(getDistance(p2, p4))
        list.add(getDistance(p3, p4))
        
        val m = list.groupingBy { it }.eachCount()
        
        if(m.size != 2) return false
        if(m.contains(0)) return false
        
        return m.filter{it.value == 4}.isNotEmpty()
    }
    
    fun getDistance(a: IntArray, b: IntArray): Int {
        val x = a[0] - b[0]
        val y = a[1] - b[1]
        return x * x + y * y
    }
}